crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,320 @@
+use crate::ports::P2pDb;
+use dashmap::DashMap;
+use fuel_core_metrics::p2p_metrics::{
+    increment_p2p_req_res_cache_hits,
+    increment_p2p_req_res_cache_misses,
+};
+use fuel_core_storage::Result as StorageResult;
+use fuel_core_types::{
+    blockchain::SealedBlockHeader,
+    services::p2p::Transactions,
+};
+use std::ops::Range;
+type BlockHeight = u32;
+
+pub struct CachedView {
+    sealed_block_headers: DashMap<BlockHeight, SealedBlockHeader>,
+    transactions_on_blocks: DashMap<BlockHeight, Transactions>,
+    metrics: bool,
+}
+
+impl CachedView {
+    pub fn new(metrics: bool) -> Self {
+        Self {
+            sealed_block_headers: DashMap::new(),
+            transactions_on_blocks: DashMap::new(),
+            metrics,
+        }
+    }
+
+    pub fn clear(&self) {
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    fn get_from_cache_or_db<V, T, F>(
+        &self,
+        cache: &DashMap<u32, T>,
+        view: &V,
+        range: Range<u32>,
+        fetch_fn: F,
+    ) -> StorageResult<Option<Vec<T>>>
+    where
+        V: P2pDb,
+        T: Clone,
+        F: Fn(&V, Range<u32>) -> StorageResult<Option<Vec<T>>>,
+    {
+        let mut items = Vec::new();
+        let mut missing_start = None;
+
+        for height in range.clone() {
```

Comment:
I think you can go around it by copying the range bounds before consuming the range itself, since those are `Copy`.  

e.g. 

```rust 
let range_copy = range.start..range.end
for height in range { /* ... */ } // No need to clone anymore
```

In this case you only need `range.end`, so it might be better to copy that value only